tests/unit/fem/test_lin_interp.cpp

```diff
@@ -59,164 +96,1226 @@ double FdotG3(const Vector & x)
+         }
+      }
+      SECTION("Operators on H1^d for element type " + std::to_string(type))
+      {
+         H1_FECollection    fec_h1(order_h1, dim);
+         FiniteElementSpace fespace_h1(&mesh, &fec_h1, dim);
+
+         GridFunction f0(&fespace_h1);
+         f0.ProjectCoefficient(FCoef);
+
+         SECTION("Mapping to HCurl")
+         {
+            ND_FECollection    fec_ndp(order_nd, dim);
+            FiniteElementSpace fespace_ndp(&mesh, &fec_ndp);
+
+            GridFunction f1(&fespace_ndp);
+
+            DiscreteLinearOperator Op(&fespace_h1,&fespace_ndp);
+            Op.AddDomainInterpolator(new IdentityInterpolator());
+            Op.Assemble();
+
+            Op.Mult(f0,f1);
+
+            REQUIRE( f1.ComputeL2Error(FCoef) < tol );
+         }
+         SECTION("Mapping to HDiv")
+         {
+            RT_FECollection    fec_rtp(order_rt, dim);
+            FiniteElementSpace fespace_rtp(&mesh, &fec_rtp);
+
+            GridFunction f2(&fespace_rtp);
+
+            DiscreteLinearOperator Op(&fespace_h1,&fespace_rtp);
+            Op.AddDomainInterpolator(new IdentityInterpolator());
+            Op.Assemble();
+
+            Op.Mult(f0,f2);
+
+            REQUIRE( f2.ComputeL2Error(FCoef) < tol );
+         }
+         /// The following tests would fail.  The reason for the
+         /// failure would not be obvious from the user's point of
+         /// view.  I recommend keeping these tests here as a reminder
+         /// that we should consider supporting this, or a very
+         /// similar, usage.
+         /*
+              SECTION("Mapping to L2^d")
+              {
+                 L2_FECollection    fec_l2(order_l2, dim);
+                 FiniteElementSpace fespace_l2(&mesh, &fec_l2, dim);
+
+                 GridFunction f2d(&fespace_l2);
+
+                 DiscreteLinearOperator Op(&fespace_h1,&fespace_l2);
+                 Op.AddDomainInterpolator(new IdentityInterpolator());
+                 Op.Assemble();
+
+                 Op.Mult(f0,f2d);
+
+                 REQUIRE( f2d.ComputeL2Error(FCoef) < tol );
+              }
+              SECTION("Mapping to L2^d (INTEGRAL)")
+              {
+                 L2_FECollection    fec_l2(order_l2, dim,
+                                           BasisType::GaussLegendre,
+                                           FiniteElement::INTEGRAL);
+                 FiniteElementSpace fespace_l2(&mesh, &fec_l2, dim);
+
+                 GridFunction f2d(&fespace_l2);
+
+                 DiscreteLinearOperator Op(&fespace_h1,&fespace_l2);
+                 Op.AddDomainInterpolator(new IdentityInterpolator());
+                 Op.Assemble();
+
+                 Op.Mult(f0,f2d);
+
+                 REQUIRE( f2d.ComputeL2Error(FCoef) < tol );
+              }
+         */
+      }
+   }
+}
+
+TEST_CASE("3D Identity Linear Interpolators",
+          "[IdentityInterpolator]")
+{
+   int order_h1 = 1, order_nd = 2, order_rt = 1, order_l2 = 1, n = 3, dim = 3;
```

Comment:
It looks like there is a lot of code duplication between this case and the 2D one. The main differences seem to be the declaration of `mesh` and the loop over elements. It is common in unit tests to have a loop over dimensions and a switch for the `mesh` declaration. Specifying the element loop in 2D and 3D could be done by setting the lower and upper bounds for the loop as variables. It would also be possible to put this in a function called by the `TEST_CASE` for each dimension and element type (one test case for 2D and 3D). Optionally, if you want to be really fancy, you could use `GENERATE`, cf. test_pa_kernels.cpp.

Reply:
I see @dylan-copeland 's point, the test code could be a bit shorter, but since these are just unit tests, though longer, they are easily readable. I am fine either way.

Reply:
I agree with you, @dylan-copeland, I had considered performing a loop over the dimension but the function coefficients stopped me.  In most cases different coefficients are needed in different dimensions.  This could be solved in some cases by making the functions aware of the dimension, replacing some coefficients by pointers to coefficients (and pointing to different things based on dimension), or using conditionals when using coefficients in yet other cases.  I chose not to proceed with these approaches so that I could keep the flow of these tests easier to follow.

Clearly there are competing priorities here and perhaps I've let these get too long.  I'd be happy to try for a compromise if you like.

Reply:
@mlstowell I did not notice the difference in function coefficients. As you mentioned, this could be simplified by making the functions aware of the dimension. This is done, for example, in test_pa_coeff.cpp. It seems like a simple change to greatly reduce the amount of code. On one hand, it may be a little easier to follow with separate 2D and 3D versions, but on the other hand it is easier to maintain and update the tests in the future if code is not duplicated, so I am in favor of reducing to one test for 2D/3D with loops and coefficient functions that switch based on the dimension.

Reply:
This is a good point.  Coefficients are more of a hassle when derivatives and products with material coefficients are involved.  For these tests the coefficients should be much easier to handle.  I'll take a look at collapsing these tests into loops.

Reply:
Collapsing into a loop cut nearly 400 lines of code and I don't think it's any more difficult to follow.